Sources_v3/Workers/EventNotificationCenter.swift

```diff
@@ -0,0 +1,37 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+class EventNotificationCenter {
+    let notificationCenter: NotificationCenter
+    let middlewares: [EventMiddleware]
+    
+    init(notificationCenter: NotificationCenter, middlewares: [EventMiddleware]) {
+        self.notificationCenter = notificationCenter
+        self.middlewares = middlewares
+    }
+    
+    func publish(_ event: Event, completion: ((Error?) -> Void)? = nil) {
+        middlewares.process(event: event) { [weak self] in
+            guard let self = self else { return }
+            
+            if let processedEvent = $0 {
+                let notification = Notification(newEventReceived: processedEvent, sender: self)
+                self.notificationCenter.post(notification)
+                completion?(nil)
+            } else {
+                completion?(ClientError.EventProcessingFailed(event: event))
```

Comment:
It doesn't mean failure, it can just mean, the event is private, and shouldn't be published. I'm not sure if we even need this completion to be honest.

Reply:
I was thinking about the case when we wanna publish multiple events (that we get from `/sync` endpoint, for example)

Won't it be better to wait until the current event is processed before starting with the next one?

Reply:
I see. This hopefully won't be needed, as the order of the events doesn't really matter and the whole middleware pre-processing is built the way it can (should) be possible to run in parallel.

Reply:
For example `database.write { }` happens on one serial queue. So if you give it a hundred events it should process them one by one in the order you provide them.

Reply:
understood. The completion is not there anymore

Reply:
done in 9ccf3e7d5bc8ffbd0021704e53abe9deed9320bb